src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs

```diff
@@ -27,90 +29,102 @@ protected SwitchBinder(Binder next, SwitchStatementSyntax switchSyntax)
 
         internal static SwitchBinder Create(Binder next, SwitchStatementSyntax switchSyntax)
         {
-            // When pattern-matching is enabled, we use a completely different binder and binding
-            // strategy for switch statements. Once we have confirmed that it is totally upward
-            // compatible with the existing syntax and semantics, we will remove *this* binder
-            // and use the new one for binding all switch statements.
             var parseOptions = switchSyntax?.SyntaxTree?.Options as CSharpParseOptions;
             return
-                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false &&
-                (parseOptions?.Features.ContainsKey("typeswitch") != false || IsPatternSwitch(switchSyntax)))
+                // in C# 6 and earlier, we use the old binder. In C# 7 and later, we use the new binder which
+                // is capable of binding both the old and new syntax. However, the new binder does not yet
+                // lead to a translation that fully supports edit-and-continue.
+                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false || parseOptions?.Features.ContainsKey("typeswitch") != false)
```

Comment:
Straight-line C# 7 code *can* hit the code path. However, it would require writing new tests for each scenario. We instead run the existing C# 6 tests under this feature flag to get the test coverage without having to write entirely new tests.

---
In reply to: [70192167](https://github.com/dotnet/roslyn/pull/12416#discussion_r70192167) [](ancestors = 70192167)